### ISSUE TITLE
[Question/Feedback]: use of submodules for child resources

### ISSUE THREAD

**Issue by kewalaka** [Azure/Azure-Verified-Modules#1457]
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Description

Looking for guidance on whether child resources should generally be a submodule.  I see this used in some of the more complex modules and it makes sense.

I've also seen it used in this proposal where it is more of a wrapper

https://github.com/Azure/terraform-azurerm-avm-res-communication-emailservice/tree/krbar/initialModuleVersion

I understand submodules provide potential for role separation , i.e. in theory a child resource could be made for an existing parent.

Was wondering if submodules for child resources are preferred or what criteria should guide their use.

Question came up here:

https://github.com/Azure/terraform-azurerm-avm-res-dbforpostgresql-flexibleserver/issues/8

**Comment by gnfsdybvik** [comment]
I was about to create a new issue, but found this which I think is asking for the same. I still want to elaborate my problem.

In Azure there are quite a few resources that are container resources for other resources or content. Like Sql Server, Storage Accounts, Managed Environment, Api Management, App Configuration etc. (there are more).
Some of these are expensive resources which you do not want to add for each service you create in your application. And you also want to reduce the number of resources to manage, E.g. you have one Sql Server instance where you keep the databases for all your services in your system. 
So in our system we want to deploy the Sql Server in a bicep script for common resources (system.bicep ). Then each service will run their own bicep script (service.bicep). The services which needs a database will add that database to the existing Sql Server resource. But the way the Sql Server module is constructed now, you have to know all the databases up front. There is no module for a database ('br/public:avm/res/sql/database'), only for the server. 
In Container Apps this is implemented better, as the app managed environment and container app are separated in different modules.

My solution for now is that I have to create my own modules (or use az cli in the pipeline) for adding resources or content into my common resources. I have that for adding a Sql Database, adding values to App Configuration, adding Apis (and more) to Api Management. For blob storage I have to reside to adding the blob containers up front because I haven't found a way to add that afterwards.

**Comment by crespo21** [comment]
Have you tried azd?

**Comment by matt-FFFFFF** [comment]
We are going to produce guidance on them prior to v1..

But yes child resources should be sub modules, this allows them to be called independently.

Any dependencies (e.g. parent resource id) should be passed as a string of the full resource id

**Comment by matt-FFFFFF** [comment]
> I also think the guidance here should take into account the complexity of deployments created by this pattern, as exemplified in [#1237](https://github.com/Azure/Azure-Verified-Modules/issues/1237).

Good point - though this is only an issue for Bicep, as Terraform does not use deployments

**Comment by AlexanderSehr** [comment]
To briefly add to the subject (on the Bicep side):
The feature to deploy child-modules directly has actually been implemented in (Bicep)-AVMs-Source repository, that is CARML. The only reason this has not been ported to the Public Registry yet is due to additional manual steps required for new modules (which includes child-modules) to be published - and given that child-modules are numerous and could be added to a module at any time, we had to figure a workaround and have been working on this for a while. Right now we're actually working on a PoC - so fingers crossed that we can enable this feature soon.
Other references: 
- https://github.com/Azure/bicep-registry-modules/issues/2868 
- https://github.com/Azure/bicep-registry-modules/issues/1933 
- https://github.com/Azure/bicep-registry-modules/issues/1390 
- https://github.com/Azure/bicep-registry-modules/issues/2899 
- https://github.com/Azure/Azure-Verified-Modules/issues/578 
- https://github.com/Azure/Azure-Verified-Modules/issues/747
-  https://github.com/Azure/bicep-registry-modules/issues/3894#issuecomment-2573257692

**Comment by brettmillerb** [comment]
@AlexanderSehr shared an update on the child modules POC on the Feb 25 community call:

https://youtu.be/M2dJetgK2U8?si=2aGGz_CMNgfws9CM&t=890

**Comment by sinedied** [comment]
@AlexanderSehr any updates regarding GA for child modules? Is there maybe a way to ask for specific submodules to be included in the pilot (I'm currently stuck with App Service app settings, had to copy the module locally which isn't great...)?

Thanks!

**Comment by sinedied** [comment]
Thanks for the update, I'll be following closely your PR to avoid future issues :)

**Comment by AlexanderSehr** [comment]
Hey @sinedied,
just a quick update in that I 'finished' the PR. As expected it turned out to be quite a big one, and 'upgrading' to it will require a careful look at the changes done to the `tests/e2e` folder changes - but I'm quite happy with the result. Once this PR and one of its dependencies is merged, we could (in alignment with the module owners @tsc-buddy & @pankajagrawal16) publish the `config` module as a child module. Ideally we do the same for `slots` as it contains the interfaces for the `config`, allowing you to even import them to your deployment using something along the lines of `import { configType } from 'br/public:avm/res/web/site/slot:0.16.0'`.

**Comment by ahmedgabers** [comment]
Hello AVM team, thank you for the excellent work on the modules.

I'm posting here since the main topic pertains to the same issue. I just wanted to clarify the design philosophy for structuring the Terraform modules, specifically when to use a flat structure vs nested submodules for child resources?

I've noticed a difference in patterns across the modules:

*   **Flat Structure (in root):**
    *   [AKS](https://github.com/Azure/terraform-azurerm-avm-ptn-aks-production/blob/main/main.tf) for `node_pools`
    *   [Front Door](https://github.com/Azure/terraform-azurerm-avm-res-cdn-profile) for `origin_groups`, `routes`, ..etc

*   **Submodule Structure (`./modules`):**
    *   [Key Vault](https://github.com/Azure/terraform-azurerm-avm-res-keyvault-vault/tree/main/modules) for `secrets` and `keys`
    *   [Virtual Network](https://github.com/Azure/terraform-azurerm-avm-res-network-virtualnetwork/tree/main/modules) for `subnets` and `peerings`

What are the key differentiators that guide this decision?

It seems all these child resources are management plane resources descending from the parent, so I am not sure if there's a correlation with the Azure API plane, which makes the pattern a bit confusing to me. For example - they all seem to warrant nested submodules:
*   `Microsoft.KeyVault/vaults/secrets`
*   `Microsoft.ContainerService/managedClusters/agentPools`
*   `Microsoft.Network/virtualNetworks/subnets`
*  `Microsoft.Cdn/profiles/endpoints`

But AKS/CDN authors chose otherwise. Any insight into the guiding principles here would be super helpful. Thanks.

**Comment by AlexanderSehr** [comment]
> Hello AVM team, thank you for the excellent work on the modules.
> 
> I'm posting here since the main topic pertains to the same issue. I just wanted to clarify the design philosophy for structuring the Terraform modules, specifically when to use a flat structure vs nested submodules for child resources?
> 
> I've noticed a difference in patterns across the modules:
> 
> * **Flat Structure (in root):**
>   
>   * [AKS](https://github.com/Azure/terraform-azurerm-avm-ptn-aks-production/blob/main/main.tf) for `node_pools`
>   * [Front Door](https://github.com/Azure/terraform-azurerm-avm-res-cdn-profile) for `origin_groups`, `routes`, ..etc
> * **Submodule Structure (`./modules`):**
>   
>   * [Key Vault](https://github.com/Azure/terraform-azurerm-avm-res-keyvault-vault/tree/main/modules) for `secrets` and `keys`
>   * [Virtual Network](https://github.com/Azure/terraform-azurerm-avm-res-network-virtualnetwork/tree/main/modules) for `subnets` and `peerings`
> 
> What are the key differentiators that guide this decision?
> 
> It seems all these child resources are management plane resources descending from the parent, so I am not sure if there's a correlation with the Azure API plane, which makes the pattern a bit confusing to me. For example - they all seem to warrant nested submodules:
> 
> * `Microsoft.KeyVault/vaults/secrets`
> * `Microsoft.ContainerService/managedClusters/agentPools`
> * `Microsoft.Network/virtualNetworks/subnets`
> * `Microsoft.Cdn/profiles/endpoints`
> 
> But AKS/CDN authors chose otherwise. Any insight into the guiding principles here would be super helpful. Thanks.

Hey @matt-FFFFFF, could you speak to the TF design? 💪

**Comment by matt-FFFFFF** [comment]
@AlexanderSehr should follow the resource hierarchy for resource modules. We will mandate for v1

**Comment by sinedied** [comment]
Just wanted to give a quick update, as I've updated an infra to using child modules today with minimal changes, and everything worked on the first try. Great job here 🎉
